Flaky Test: TestNoControlConnWhenDown sometimes fails because our daemon holds an HTTP request open after being stopped

From this test run: https://github.com/tailscale/tailscale/actions/runs/26246871932/job/77248184036?pr=19658#step:9:462

```
=== RUN   TestNoControlConnWhenDown
=== PAUSE TestNoControlConnWhenDown
=== CONT  TestNoControlConnWhenDown
    integration.go:543: DERP httpsrv listener: 127.0.0.1:46175
    integration.go:577: control URL: http://127.0.0.1:40631/
    integration.go:625: 2026/05/21 19:19:15 DEBUG-ADDR=127.0.0.1:39871
    integration_test.go:1067: Status: NeedsLogin
    integration_test.go:1070: Running /tmp/TestNoControlConnWhenDown855130255/001/tailscale --socket=/tmp/TestNoControlConnWhenDown855130255/002/tailscale.sock up --login-server=http://127.0.0.1:40631/ --reset ...
    logger.go:105: testcontrol: Created user userid:1 for node nodekey:a7fbba3c63838154158e971941e7ae3024bf478a20aa92f6502e3d7b3db5fa21
    integration.go:918: Running down ...
    integration.go:625: 2026/05/21 19:19:16 DEBUG-ADDR=127.0.0.1:42231
    integration_test.go:1080: Status: Stopped
    integration_test.go:1091: in serve map = 1; want 0
    stuntest.go:65: STUN server shutdown
--- FAIL: TestNoControlConnWhenDown (0.59s)
ok	tailscale.com/version	3.806s
ok	tailscale.com/version/distro	(cached)
ok	tailscale.com/version/mkversion	(cached)
FAIL
FAIL	tailscale.com/tstest/integration	29.140s
```

The failure happens here: https://github.com/tailscale/tailscale/blob/fba431f825d170ba6a8faa78ca26e7776c45eebf/tstest/integration/integration_test.go#L1089-L1092

Claude Code that this can be reliably triggered by slowing down the mock control plane after its request context has been cancelled:

https://github.com/tailscale/tailscale/blob/7dabebc6913511c9aeca251a34e35ad95e357488/tstest/integration/testcontrol/testcontrol.go#L1323-L1348

``` diff
diff --git i/tstest/integration/testcontrol/testcontrol.go w/tstest/integration/testcontrol/testcontrol.go
index d02eef5f4..7223de947 100644
--- i/tstest/integration/testcontrol/testcontrol.go
+++ w/tstest/integration/testcontrol/testcontrol.go
@@ -1333,6 +1333,7 @@ func (s *Server) serveMap(w http.ResponseWriter, r *http.Request, mkey key.Machi
 				if keepAliveTimer != nil {
 					keepAliveTimer.Stop()
 				}
+				time.Sleep(time.Second)
 				return
 			case _, ok := <-updatesCh:
 				if !ok {
```

Interestingly, this causes the test to fail even harder, with 2 connections in serve map:
```
--- FAIL: TestNoControlConnWhenDown (3.38s)
    flakytest.go:62: flakytest: issue tracking this flaky test: https://github.com/tailscale/tailscale/issues/19831
    integration.go:543: DERP httpsrv listener: 127.0.0.1:39315
    integration.go:577: control URL: http://127.0.0.1:39413
    integration.go:625: 2026/05/21 16:48:53 DEBUG-ADDR=127.0.0.1:39493
    integration_test.go:1069: Status: NeedsLogin
    integration_test.go:1072: Running /tmp/TestNoControlConnWhenDown799498999/001/tailscale --socket=/tmp/TestNoControlConnWhenDown799498999/002/tailscale.sock up --login-server=http://127.0.0.1:39413 --reset ...
    logger.go:105: testcontrol: Created user userid:1 for node nodekey:fce9ea2711ce31b559b0f2c7f6b9cdc801fd5d8310761e5534041672cfc2794c
    integration.go:918: Running down ...
    integration.go:625: 2026/05/21 16:48:53 DEBUG-ADDR=127.0.0.1:45151
    integration_test.go:1082: Status: Stopped
    integration_test.go:1093: in serve map = 2; want 0
    stuntest.go:65: STUN server shutdown
FAIL
FAIL	tailscale.com/tstest/integration	3.392s
FAIL
```

--- a/ipn/ipnlocal/local.go
+++ b/ipn/ipnlocal/local.go
@@ -1037,13 +1037,37 @@ func (b *LocalBackend) pauseOrResumeControlClientLocked() {
 	if b.cc == nil {
 		return
 	}
-	networkUp := b.interfaceState.AnyInterfaceUp()
-	pauseForNetwork := (b.state == ipn.Stopped && b.NetMapNoPeers() != nil) || (!networkUp && !testenv.InTest() && !envknob.AssumeNetworkUp())
+	b.cc.SetPaused(b.shouldPauseControlClientLocked(b.pm.CurrentPrefs()))
+}
 
-	prefs := b.pm.CurrentPrefs()
+// shouldPauseControlClientLocked reports whether the control client should be paused:
+// if the LocalBackend is in Stopped state with a valid NetMap,
+// if there is no network available,
+// or if syncing preferences from the control plane has been disabled.
+//
+// b.mu must be held.
+func (b *LocalBackend) shouldPauseControlClientLocked(prefs ipn.PrefsView) bool {
+	syncs.RequiresMutex(&b.mu)
+
 	pauseForSyncPref := prefs.Valid() && prefs.Sync().EqualBool(false)
+	if pauseForSyncPref {
+		return true
+	}
 
-	b.cc.SetPaused(pauseForNetwork || pauseForSyncPref)
+	// If tailscaled is being restarted, but it is supposed to be stopped,
+	// it mustn’t pause until the initial netmap has been loaded.
+	isStopped := b.state == ipn.Stopped && b.NetMapNoPeers() != nil
+	if isStopped {
+		return true
+	}
+
+	networkUp := b.interfaceState.AnyInterfaceUp()
+	pauseForNetwork := !networkUp && !testenv.InTest() && !envknob.AssumeNetworkUp()
+	if pauseForNetwork {
+		return true
+	}
+
+	return false
 }
 
 // DisconnectControl shuts down control client. This can be run before node shutdown to force control to consider this ndoe
@@ -2961,7 +2985,7 @@ func (b *LocalBackend) startLocked(opts ipn.Options) error {
 		ControlKnobs:         b.sys.ControlKnobs(),
 		Shutdown:             ccShutdown,
 		Bus:                  b.sys.Bus.Get(),
-		StartPaused:          prefs.Sync().EqualBool(false),
+		StartPaused:          b.shouldPauseControlClientLocked(prefs),
 	})
 	if err != nil {
 		return err

--- a/tstest/integration/integration.go
+++ b/tstest/integration/integration.go
@@ -919,6 +919,20 @@ func (n *TestNode) MustDown() {
 	if err := n.Tailscale("down", "--accept-risk=all").Run(); err != nil {
 		t.Fatalf("down: %v", err)
 	}
+
+	// The tailscale down command is asynchronous, so it returns early.
+	// Wait for tailscaled to drop its connection before continuing.
+	if err := tstest.WaitFor(time.Second, func() error {
+		if err := t.Context().Err(); err != nil {
+			return err
+		}
+		if c := n.env.Control.InServeMap(); c != 0 {
+			return fmt.Errorf("%d connections remaining in serve map", c)
+		}
+		return nil
+	}); err != nil {
+		t.Fatalf("tailscale down: %v", err)
+	}
 }
 
 func (n *TestNode) MustLogOut() {

--- a/tstest/integration/integration_test.go
+++ b/tstest/integration/integration_test.go
@@ -1058,8 +1058,6 @@ func TestC2NPingRequest(t *testing.T) {
 // Issue 2434: when "down" (WantRunning false), tailscaled shouldn't
 // be connected to control.
 func TestNoControlConnWhenDown(t *testing.T) {
-	flakytest.Mark(t, "https://github.com/tailscale/tailscale/issues/19831")
-
 	tstest.Shard(t)
 	tstest.Parallel(t)
 	env := NewTestEnv(t)
@@ -1083,14 +1081,24 @@ func TestNoControlConnWhenDown(t *testing.T) {
 
 	n1.AwaitBackendState("Stopped")
 
+	// The real test: verify our daemon doesn't have an HTTP request open.
+	// Stopping the client may take some time to disconnect from testcontrol.
+	if err := tstest.WaitFor(time.Second, func() error {
+		if n := env.Control.InServeMap(); n != 0 {
+			return fmt.Errorf("in serve map = %d; want 0", n)
+		}
+		return nil
+	}); err != nil {
+		t.Fatalf("unexpected connections while stopped: %v", err)
+	}
+
 	ip2 := n1.AwaitIP4()
 	if ip1 != ip2 {
 		t.Errorf("IPs different: %q vs %q", ip1, ip2)
 	}
 
-	// The real test: verify our daemon doesn't have an HTTP request open.
 	if n := env.Control.InServeMap(); n != 0 {
-		t.Errorf("in serve map = %d; want 0", n)
+		t.Fatalf("unexpected connection triggered by tailscale ip: in serve map = %d; want 0", n)
 	}
 
 	d2.MustCleanShutdown(t)

